File: sliprun/api/slipstream.py
"""
Marathon Slipstream API client.

Endpoints:
  GET  /api/system                  - chain info, block height, fee floor
  GET  /api/block-info              - structured blockchain status
  GET  /api/rates                   - current fee rates (sat/vByte)
  GET  /api/transactions/status     - transaction status by txid
  POST /api/transactions            - submit single raw transaction
  POST /api/transactions/packages   - submit 2-25 transactions as a package
  POST /api/mempool/tests           - test transactions against consensus rules
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import requests


class SlipstreamError(Exception):
    def __init__(self, status_code: int, body: str):
        self.status_code = status_code
        self.body = body
        super().__init__(f"Slipstream API error {status_code}: {body}")


@dataclass
class SystemInfo:
    version: str
    chain: str
    block_height: int
    fee_rate_floor: float

    @classmethod
    def from_dict(cls, d: dict) -> "SystemInfo":
        return cls(
            version=d.get("version", ""),
            chain=d.get("chain_name", d.get("chain", "")),
            block_height=d.get("block_height", 0),
            fee_rate_floor=d.get("fee_rate_floor", 0.0),
        )


@dataclass
class FeeRates:
    low: float
    medium: float
    high: float
    unit: str = "sat/vByte"

    @classmethod
    def from_dict(cls, d: dict) -> "FeeRates":
        # Slipstream returns rates in various shapes; normalise here
        return cls(
            low=d.get("low", d.get("economy", 0.0)),
            medium=d.get("medium", d.get("normal", 0.0)),
            high=d.get("high", d.get("priority", 0.0)),
        )


@dataclass
class TxSubmitResult:
    txid: str
    status: str
    message: str

    @classmethod
    def from_dict(cls, d: dict) -> "TxSubmitResult":
        return cls(
            txid=d.get("txid", d.get("tx_id", "")),
            status=d.get("status", ""),
            message=d.get("message", ""),
        )


class SlipstreamClient:
    """HTTP client for the Marathon Slipstream API."""

    BASE_URL = "https://slipstream.mara.com"

    def __init__(
        self,
        base_url: str = BASE_URL,
        client_code: str | None = None,
        timeout: int = 30,
    ):
        self.base_url = base_url.rstrip("/")
        self.client_code = client_code
        self._timeout = timeout
        self._session = requests.Session()
        self._session.headers.update(
            {"Content-Type": "application/json", "Accept": "application/json"}
        )

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _get(self, path: str, params: dict[str, Any] | None = None) -> Any:
        url = f"{self.base_url}{path}"
        resp = self._session.get(url, params=params or {}, timeout=self._timeout)
        if not resp.ok:
            raise SlipstreamError(resp.status_code, resp.text)
        return resp.json()

    def _post(self, path: str, payload: dict) -> Any:
        url = f"{self.base_url}{path}"
        resp = self._session.post(url, json=payload, timeout=self._timeout)
        if not resp.ok:
            raise SlipstreamError(resp.status_code, resp.text)
        return resp.json()

    def _with_client_code(self, d: dict) -> dict:
        if self.client_code:
            d["client_code"] = self.client_code
        return d

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get_system_info(self) -> dict:
        """Returns API version, chain name, block height, and fee rate floor."""
        return self._get("/api/system")

    def get_block_info(self) -> dict:
        """Returns structured blockchain status data."""
        return self._get("/api/block-info")

    def get_rates(self) -> dict:
        """Returns current fee rates in sat/vByte."""
        params: dict[str, Any] = {}
        if self.client_code:
            params["client_code"] = self.client_code
        return self._get("/api/rates", params)

    def get_transaction_status(self, tx_id: str) -> dict:
        """Returns detailed status for a transaction by its txid."""
        return self._get("/api/transactions/status", {"tx_id": tx_id})

    def submit_transaction(self, tx_hex: str) -> dict:
        """
        Submit a single raw Bitcoin transaction.

        Args:
            tx_hex: Signed transaction in hexadecimal format.

        Returns:
            API response with status and txid.
        """
        payload = self._with_client_code({"tx_hex": tx_hex})
        return self._post("/api/transactions", payload)

    def submit_package(self, tx_hexes: list[str]) -> dict:
        """
        Submit a package of 2-25 related transactions (e.g. commit + reveal).

        The API processes them as a unit, which is required for CPFP / Ordinals
        commit-reveal pairs where the reveal depends on the unconfirmed commit.

        Args:
            tx_hexes: Ordered list of raw transaction hexes (2-25 items).

        Returns:
            API response mapping txids to per-transaction results.
        """
        if not (2 <= len(tx_hexes) <= 25):
            raise ValueError(
                f"Package must contain 2-25 transactions, got {len(tx_hexes)}"
            )
        payload = self._with_client_code({"tx_hexes": tx_hexes})
        return self._post("/api/transactions/packages", payload)

    def test_transaction(self, tx_hexes: list[str]) -> dict:
        """
        Dry-run: test transactions against consensus and policy rules without
        broadcasting.

        Args:
            tx_hexes: List of raw transaction hexes to validate.

        Returns:
            Validation results per transaction.
        """
        return self._post("/api/mempool/tests", {"tx_hexes": tx_hexes})

File: web/src/lib/slipstream.ts
// Slipstream API client — ported from sliprun/api/slipstream.py
// Mainnet: slipstream.mara.com  Testnet: teststream.mara.com

export class SlipstreamError extends Error {
  constructor(
    public readonly status: number,
    public readonly body: string
  ) {
    super(`Slipstream API error ${status}: ${body}`)
    this.name = 'SlipstreamError'
  }
}

function resolveUrl(path: string, network: 'mainnet' | 'testnet' | 'testnet4'): string {
  if (network === 'testnet' || network === 'testnet4') {
    const base = import.meta.env.VITE_TESTSTREAM_URL ?? ''
    // path is e.g. /api/rates — rewrite to /testapi/rates for Vite proxy
    return base.startsWith('http') ? `${base}${path}` : path.replace(/^\/api/, '/testapi')
  }
  const base = import.meta.env.VITE_SLIPSTREAM_URL ?? ''
  return base.startsWith('http') ? `${base}${path}` : path
}

async function apiFetch(path: string, network: 'mainnet' | 'testnet' | 'testnet4' = 'mainnet', init?: RequestInit): Promise<unknown> {
  const url = resolveUrl(path, network)
  const resp = await fetch(url, {
    headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
    ...init,
  })

  if (!resp.ok) {
    const body = await resp.text().catch(() => '')
    throw new SlipstreamError(resp.status, body)
  }

  return resp.json()
}

export interface FeeRates {
  low: number
  medium: number
  high: number
}

export async function getRates(clientCode?: string, network: 'mainnet' | 'testnet' | 'testnet4' = 'mainnet'): Promise<FeeRates> {
  const params = clientCode ? `?client_code=${encodeURIComponent(clientCode)}` : ''
  const data = (await apiFetch(`/api/rates${params}`, network)) as Record<string, number>
  // Slipstream returns: submit_fee_rate, market_rate, slipstream_rate (effective_rate)
  return {
    low: data.submit_fee_rate ?? data.low ?? data.economy ?? 2,
    medium: data.market_rate ?? data.medium ?? data.normal ?? 5,
    high: data.slipstream_rate ?? data.effective_rate ?? data.high ?? data.priority ?? 10,
  }
}

export interface TxStatus {
  txid?: string
  status?: string
  message?: string
  [key: string]: unknown
}

export async function getTransactionStatus(txId: string, network: 'mainnet' | 'testnet' | 'testnet4' = 'mainnet'): Promise<TxStatus> {
  return apiFetch(`/api/transactions/status?tx_id=${encodeURIComponent(txId)}`, network) as Promise<TxStatus>
}

export interface PackageResult {
  [txid: string]: unknown
}

export async function submitPackage(
  txHexes: [string, string],
  clientCode?: string,
  network: 'mainnet' | 'testnet' | 'testnet4' = 'mainnet'
): Promise<PackageResult> {
  const payload: Record<string, unknown> = { tx_hexes: txHexes }
  if (clientCode) payload.client_code = clientCode
  return apiFetch('/api/transactions/packages', network, {
    method: 'POST',
    body: JSON.stringify(payload),
  }) as Promise<PackageResult>
}

export async function testTransactions(txHexes: string[]): Promise<unknown> {
  return apiFetch('/api/mempool/tests', {
    method: 'POST',
    body: JSON.stringify({ tx_hexes: txHexes }),
  })
}
